libp2p/UDP.h

```diff
@@ -0,0 +1,266 @@
+	static Public authenticate(bytesConstRef _sig, bytesConstRef _rlp);
+
+	RLPXDatagramFace(bi::udp::endpoint const& _ep): UDPDatagram(_ep) {}
+	virtual h256 sign(Secret const& _from);
+
+	virtual void streamRLP(RLPStream&) const =0;
+	virtual void interpretRLP(bytesConstRef _bytes) =0;
+};
+
+template <class T>
+struct RLPXDatagram: public RLPXDatagramFace
+{
+	using RLPXDatagramFace::RLPXDatagramFace;
+	static T fromBytesConstRef(bi::udp::endpoint const& _ep, bytesConstRef _bytes) { T t(_ep); t.interpretRLP(_bytes); return std::move(t); }
+};
+
+/**
+ * @brief Interface which UDPSocket will implement.
+ */
+struct UDPSocketFace
+{
+	virtual bool send(UDPDatagram const& _msg) = 0;
+	virtual void disconnect() = 0;
+};
+
+/**
+ * @brief Interface which a UDPSocket's owner must implement.
+ */
+struct UDPSocketEvents
+{
+	virtual void onDisconnected(UDPSocketFace*) {};
+	virtual void onReceived(UDPSocketFace*, bi::udp::endpoint const& _from, bytesConstRef _packetData) = 0;
+};
+	
+/**
+ * @brief UDP Interface
+ * Handler must implement UDPSocketEvents.
+ *
+ * @todo multiple endpoints (we cannot advertise 0.0.0.0)
+ * @todo decouple deque from UDPDatagram and add ref() to datagram for fire&forget
+ */
+template <typename Handler, unsigned MaxDatagramSize>
+class UDPSocket: UDPSocketFace, public std::enable_shared_from_this<UDPSocket<Handler, MaxDatagramSize>>
+{
+public:
+	enum { maxDatagramSize = MaxDatagramSize };
+	static_assert(maxDatagramSize < 65507, "UDP datagrams cannot be larger than 65507 bytes");
+	
+	UDPSocket(ba::io_service& _io, UDPSocketEvents& _host, unsigned _port): m_host(_host), m_endpoint(bi::udp::v4(), _port), m_socket(_io) { m_started.store(false); m_closed.store(true); };
+	virtual ~UDPSocket() { disconnect(); }
+
+	/// Socket will begin listening for and delivering packets
+	void connect();
+
+	/// Send datagram.
+	bool send(UDPDatagram const& _datagram);
+	
+	/// Returns if socket is open.
+	bool isOpen() { return !m_closed; }
+
+	/// Disconnect socket.
+	void disconnect() { disconnectWithError(boost::asio::error::connection_reset); }
+	
+protected:
+	
+	void doRead();
+	
+	void doWrite();
+	
+	void disconnectWithError(boost::system::error_code _ec);
+
+	std::atomic<bool> m_started;					///< Atomically ensure connection is started once. Start cannot occur unless m_started is false. Managed by start and disconnectWithError.
+	std::atomic<bool> m_closed;					///< Connection availability.
+	
+	UDPSocketEvents& m_host;						///< Interface which owns this socket.
+	bi::udp::endpoint m_endpoint;					///< Endpoint which we listen to.
+	
+	Mutex x_sendQ;
+	std::deque<UDPDatagram> sendQ;				///< Queue for egress data.
+	std::array<byte, maxDatagramSize> recvData;	///< Buffer for ingress data.
+	bi::udp::endpoint recvEndpoint;				///< Endpoint data was received from.
```

Comment:
Thanks. At one point I had this silly idea that a UDPSocket could just be a struct :)